cmd/spire-server/cli/agent/agent_test.go

```diff
@@ -0,0 +1,290 @@
+			returnCode := test.listCmd.Run([]string{})
+			require.Contains(t, test.testEnv.Stdout.(*bytes.Buffer).String(), tt.expectedStdout)
+			require.Equal(t, tt.expectedStderr, test.testEnv.Stderr.(*bytes.Buffer).String())
+			require.Equal(t, tt.expectedReturnCode, returnCode)
+		})
+	}
+}
+
+func TestShowHelp(t *testing.T) {
+	test := setupTest(t)
+
+	test.showCmd.Help()
+	require.Equal(t, `Usage of agent show:
+  -registrationUDSPath string
+    	Registration API UDS path (default "/tmp/spire-registration.sock")
+  -spiffeID string
+    	The SPIFFE ID of the agent to show (agent identity)
+`, test.testEnv.Stderr.(*bytes.Buffer).String())
+}
+
+func TestShow(t *testing.T) {
+	for _, tt := range []struct {
+		name               string
+		args               []string
+		expectedReturnCode int
+		expectedStdout     string
+		expectedStderr     string
+		existentAgent      string
+		dsError            error
+	}{
+		{
+			name:               "success",
+			args:               []string{"-spiffeID", "spiffe://example.org/spire/agent/agent1"},
+			expectedReturnCode: 0,
+			existentAgent:      "spiffe://example.org/spire/agent/agent1",
+			expectedStdout:     "Found an attested agent given its SPIFFE ID\n\nSPIFFE ID         : spiffe://example.org/spire/agent/agent1",
+		},
+		{
+			name:               "spiffe id not found",
+			args:               []string{"-spiffeID", "spiffe://example.org/spire/agent/agent2"},
+			expectedReturnCode: 1,
+			existentAgent:      "spiffe://example.org/spire/agent/agent1",
+			expectedStderr:     "rpc error: code = NotFound desc = agent not found\n",
+		},
+		{
+			name:               "no spiffe id",
+			expectedReturnCode: 1,
+			existentAgent:      "spiffe://example.org/spire/agent/agent1",
+			expectedStderr:     "a SPIFFE ID is required\n",
+		},
+		{
+			name:               "no agents - not found",
+			expectedReturnCode: 1,
+			args:               []string{"-spiffeID", "spiffe://example.org/spire/agent/agent1"},
+			expectedStderr:     "rpc error: code = NotFound desc = agent not found\n",
+		},
+		{
+			name:               "datastore error",
+			args:               []string{"-spiffeID", "spiffe://example.org/spire/agent/agent1"},
+			expectedReturnCode: 1,
+			dsError:            errors.New("datastore error"),
+			expectedStderr:     "rpc error: code = Internal desc = failed to fetch agent: datastore error\n",
+		},
+	} {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			test := setupTest(t)
+			if tt.existentAgent != "" {
+				test.createAgent(t, tt.existentAgent)
+			}
+
+			test.ds.SetNextError(tt.dsError)
+			returnCode := test.showCmd.Run(tt.args)
+			require.Contains(t, test.testEnv.Stdout.(*bytes.Buffer).String(), tt.expectedStdout)
+			require.Equal(t, tt.expectedStderr, test.testEnv.Stderr.(*bytes.Buffer).String())
+			require.Equal(t, tt.expectedReturnCode, returnCode)
+		})
+	}
+}
+
+func setupTest(t *testing.T) *agentTest {
+	ds := fakedatastore.New(t)
+
+	service := agent.New(agent.Config{
```

Comment:
I'm a little torn here. On one hand, using the agent service implementation here provides a pretty tight test, in terms of regression coverage. On the other hand, it does introduce a very tight coupling. The fact that this unit test for a CLI component has to inject stuff into a data store is a pretty big smell to me.

I know I introduced this pattern in previous tests, but I kind of feel like this level of integration belongs in an integration test and that unit test coverage for the CLI can be more easily accomplished with a targeted,  "fake" agent service implementation.

Curious what others thoughts are here...

Reply:
We discussed this in the last maintainers call. We agreed that this implementation may be difficult to maintain and the proper place to have this kind of test is in an integration test. I'll update it to have a fake agent service implementation.